src/dbms/replication_handler.cpp

```diff
@@ -137,29 +136,36 @@ auto ReplicationHandler::RegisterReplica(const memgraph::replication::Replicatio
       case memgraph::replication::RegisterReplicaError::SUCCESS:
         break;
     }
+  }
 
   if (!allow_mt_repl && dbms_handler_.All().size() > 1) {
     spdlog::warn("Multi-tenant replication is currently not supported!");
   }
 
+#ifdef MG_ENTERPRISE
+  // Update system before enabling individual storage <-> replica clients
+  dbms_handler_.SystemRestore(*instance_client.GetValue());
+#endif
+
   bool all_clients_good = true;
 
   // Add database specific clients (NOTE Currently all databases are connected to each replica)
-  dbms_handler_.ForEach([&](Database *db) {
-    auto *storage = db->storage();
-    if (!allow_mt_repl && storage->id() != kDefaultDB) {
+  dbms_handler_.ForEach([&](DatabaseAccess db_acc) {
+    auto *storage = db_acc->storage();
+    if (!allow_mt_repl && storage->name() != kDefaultDB) {
       return;
     }
     // TODO: ATM only IN_MEMORY_TRANSACTIONAL, fix other modes
     if (storage->storage_mode_ != storage::StorageMode::IN_MEMORY_TRANSACTIONAL) return;
 
-    all_clients_good &=
-        storage->repl_storage_state_.replication_clients_.WithLock([storage, &instance_client](auto &storage_clients) {
+    all_clients_good &= storage->repl_storage_state_.replication_clients_.WithLock(
+        [storage, &instance_client, db_acc = std::move(db_acc)](auto &storage_clients) mutable {  // NOLINT
           auto client = std::make_unique<storage::ReplicationStorageClient>(*instance_client.GetValue());
-          client->Start(storage);
+          // All good, start replica client
+          client->Start(storage, std::move(db_acc));
           // After start the storage <-> replica state should be READY or RECOVERING (if correctly started)
           // MAYBE_BEHIND isn't a statement of the current state, this is the default value
-          // Failed to start due to branching of MAIN and REPLICA
+          // Failed to start due an error like branching of MAIN and REPLICA
```

Comment:
Isn't it that in the tests you allow branching? I saw test if I remember correctly that you can have replica which has different data than main because something was done prior to becoming replica and it is allowed to register that replica on main?

Reply:
This is tricky with multi-tenancy.
What this comment refers to is the storage replication client. Which handles only the replication and the replica state of one storage.
When registering we check if the storage on MAIN and REPLICA have branched. 
Now, where it gets complicated... When looking at tenants (list of all storage), we check their UUIDs and if they don't match (even if the names are the same) we handle them as different. 
From discussing on slack, we decided to force the REPLICA to follow MAIN's tenant list. Meaning if we have a tenant named "db" on MAIN and REPLICA, but the "db" on REPLICA was created separately from the current MAIN, we delete it and create a new one that completely follows the MAIN's history and so no branching.

TLDR;
System and storage replication differs in that the system is forcefully updated on REPLICAs (don't care about branching), while the storage replication cares about branching at registration time. A storage on MAIN and REPLICA are connected via the UUID, so a tenant with the same name but a different UUID is handled as a different tenant.